File: backend/services/trade_adapters/kraken.py
"""Kraken Futures linear-perp trade adapter.

Authentication (different from Kraken Spot):
  hash      = SHA256(postData + nonce + endpointPath)
  signature = HMAC-SHA512(base64_decode(secret), hash)
  authent   = base64_encode(signature)
  headers   = APIKey, Nonce, Authent

Symbol convention: PF_<TOKEN>USD (linear, USD-collateralised). XBT alias
applies for BTC. Order quantity is in CONTRACTS where 1 contract = 1
unit of the base token (no contract-multiplier complication on PF_).

Reference: https://docs.kraken.com/api/docs/futures-api/trading
"""
from __future__ import annotations

import asyncio
import base64
import hashlib
import hmac
import logging
import time
from typing import Any
from urllib.parse import urlencode

import httpx

logger = logging.getLogger("avalant.trade.kraken")

BASE = "https://futures.kraken.com"
ROOT = "/derivatives/api/v3"


def _norm_symbol(sym: str) -> str:
    sym = (sym or "").upper()
    if sym == "BTC":
        sym = "XBT"
    return f"PF_{sym}USD"


def _denorm_symbol(pf_sym: str) -> str:
    """PF_XBTUSD → BTC, PF_ETHUSD → ETH."""
    s = pf_sym
    if s.startswith("PF_"):
        s = s[3:]
    if s.endswith("USD"):
        s = s[:-3]
    if s == "XBT":
        s = "BTC"
    return s


# ── Instrument cache (size precision per market) ─────────────────────────────
_INSTR_CACHE: dict[str, dict] = {}
_INSTR_TS: float = 0.0
_INSTR_TTL = 1800.0
_INSTR_LOCK = asyncio.Lock()


async def _instruments() -> dict[str, dict]:
    global _INSTR_TS
    if _INSTR_CACHE and (time.monotonic() - _INSTR_TS) < _INSTR_TTL:
        return _INSTR_CACHE
    async with _INSTR_LOCK:
        if _INSTR_CACHE and (time.monotonic() - _INSTR_TS) < _INSTR_TTL:
            return _INSTR_CACHE
        try:
            async with httpx.AsyncClient(timeout=8) as c:
                r = await c.get(f"{BASE}{ROOT}/instruments")
                items = (r.json() or {}).get("instruments") or []
        except Exception:
            return _INSTR_CACHE
        out = {}
        for it in items:
            sym = it.get("symbol") or ""
            if not sym.startswith("PF_"):
                continue
            out[sym] = {
                "tick_size": float(it.get("tickSize") or 0),
                "contract_size": float(it.get("contractSize") or 1),
                "max_position_size": float(it.get("maxPositionSize") or 0),
                "category": it.get("category") or "",
            }
        if out:
            _INSTR_CACHE.clear()
            _INSTR_CACHE.update(out)
            _INSTR_TS = time.monotonic()
        return _INSTR_CACHE


def _sign(api_secret: str, post_data: str, nonce: str, path: str) -> str:
    """Kraken futures signature: SHA256(postData + nonce + path) then
    HMAC-SHA512 with base64-decoded secret, output base64."""
    msg = (post_data + nonce + path).encode()
    hashed = hashlib.sha256(msg).digest()
    secret_decoded = base64.b64decode(api_secret)
    sig = hmac.new(secret_decoded, hashed, hashlib.sha512).digest()
    return base64.b64encode(sig).decode()


async def _signed_request(creds: dict, method: str, path: str,
                           params: dict | None = None) -> Any:
    api_key = (creds.get("api_key") or "").strip()
    api_secret = (creds.get("api_secret") or "").strip()
    if not api_key or not api_secret:
        raise RuntimeError("Kraken Futures requires api_key and api_secret")
    post_data = urlencode(params or {})
    # Path used in the signature is RELATIVE — Kraken docs use endpoint path
    # without the /derivatives/api/v3 prefix (e.g. "/accounts").
    # The full URL still uses ROOT.
    sig_path = path
    nonce = str(int(time.time() * 1000))
    authent = _sign(api_secret, post_data, nonce, sig_path)
    headers = {
        "APIKey": api_key,
        "Nonce": nonce,
        "Authent": authent,
        "Accept": "application/json",
    }
    from backend.services.trade_adapters._http import http_client
    client = http_client(BASE, timeout=10.0)
    rel = ROOT + path
    if method == "GET":
        if post_data:
            rel += "?" + post_data
        r = await client.get(rel, headers=headers)
    else:
        headers["Content-Type"] = "application/x-www-form-urlencoded"
        r = await client.post(rel, headers=headers, content=post_data)
    if r.status_code >= 400:
        raise RuntimeError(f"Kraken {r.status_code}: {r.text[:200]}")
    data = r.json()
    if isinstance(data, dict) and data.get("result") == "error":
        raise RuntimeError(f"Kraken: {data.get('error') or data}")
    return data


class KrakenAdapter:
    EXCHANGE = "kraken"
    DISPLAY = "Kraken"

    @classmethod
    async def fetch_balance(cls, creds: dict) -> dict:
        data = await _signed_request(creds, "GET", "/accounts")
        # accounts.flex.balanceValue is the USD value of the multi-collateral
        # flex account (the standard portfolio account on Kraken Futures).
        accounts = (data or {}).get("accounts") or {}
        flex = accounts.get("flex") or {}
        usdt = float(flex.get("balanceValue") or 0)
        if usdt == 0:
            # Older accounts have a "cash" section keyed by currency — sum USD/USDT
            cash = (accounts.get("cash") or {}).get("balances") or {}
            usdt = float(cash.get("USD") or 0) + float(cash.get("USDT") or 0)
        return {"usdt": usdt}

    @classmethod
    async def validate_key(cls, creds: dict, need_trade: bool = False) -> dict:
        out = {"can_read": False, "can_trade": False, "balance_usdt": None, "error": None}

        def _friendly(e: Exception, fallback: str) -> str:
            msg = str(e)
            ml = msg.lower()
            if "authenticationerror" in ml or "401" in ml or "invalid api key" in ml:
                return "Invalid API key — check key, secret, and that it's enabled for Futures (not Spot)"
            if "signaturedidnotmatch" in ml or "signature" in ml:
                return "Signature mismatch — API secret is wrong (must be the base64 secret from the futures key page)"
            if "permission" in ml or "notauthorized" in ml or "403" in ml:
                return "API key lacks the required permission — enable General + Funding read"
            if "ipaddress" in ml or "ip" in ml and "whitelist" in ml:
                return "Server IP not whitelisted on this Kraken Futures key"
            if "ratelimited" in ml or "429" in ml:
                return "Kraken rate-limited the request — try again in a minute"
            return f"{fallback}: {msg[:140]}"

        try:
            bal = await cls.fetch_balance(creds)
            out["can_read"] = True
            out["balance_usdt"] = float(bal.get("usdt") or 0)
        except Exception as e:
            out["error"] = _friendly(e, "Kraken rejected the key")
            return out
        if need_trade:
            try:
                await _signed_request(creds, "GET", "/leveragepreferences")
                out["can_trade"] = True
            except Exception as e:
                out["error"] = _friendly(e, "Kraken trade permission missing")
        return out

    @classmethod
    async def get_public_max_leverage(cls, symbol: str) -> int:
        # Kraken Futures generally caps at 50x for majors; venue rejects above it
        return 50

    @classmethod
    async def set_leverage(cls, creds: dict, symbol: str,
                           leverage: int, margin_mode: str) -> None:
        sym = _norm_symbol(symbol)
        try:
            await _signed_request(creds, "POST", "/leveragepreferences",
                                   {"symbol": sym, "maxLeverage": int(max(1, leverage))})
        except Exception as e:
            # Kraken may reject if leverage already at the requested value;
            # not a fatal error — trade can proceed.
            logger.info("kraken set_leverage(%s, %sx) note: %s", sym, leverage, e)

    @classmethod
    async def place_order(cls, creds: dict, symbol: str, side: str, quantity: float,
                          leverage: int = 1, margin_mode: str = "isolated") -> dict:
        sym = _norm_symbol(symbol)
        is_buy = (side or "").lower() in ("buy", "long")
        params = {
            "orderType": "mkt",
            "symbol": sym,
            "side": "buy" if is_buy else "sell",
            "size": str(quantity),
        }
        data = await _signed_request(creds, "POST", "/sendorder", params)
        status = (data or {}).get("sendStatus") or {}
        order_id = status.get("order_id") or ""
        # avg_price isn't guaranteed in immediate response — pull from fills
        # if present, otherwise leave 0 and let the position sync fill it in.
        fills = status.get("fillStatus") or {}
        avg = float(fills.get("price") or 0)
        return {"order_id": str(order_id), "avg_price": avg}

    @classmethod
    async def close_position(cls, creds: dict, symbol: str, side: str) -> dict:
        positions = await cls.list_positions(creds, symbol=symbol)
        match = next((p for p in positions if (p.get("symbol") or "").upper() == symbol.upper()), None)
        if not match:
            return {"order_id": "", "closed_qty": 0.0, "realized_pnl_usd": 0.0}
        qty = abs(float(match.get("quantity") or 0))
        if qty <= 0:
            return {"order_id": "", "closed_qty": 0.0, "realized_pnl_usd": 0.0}
        opposite = "sell" if (match.get("side") or "").lower() == "buy" else "buy"
        sym = _norm_symbol(symbol)
        params = {
            "orderType": "mkt",
            "symbol": sym,
            "side": opposite,
            "size": str(qty),
            "reduceOnly": "true",
        }
        data = await _signed_request(creds, "POST", "/sendorder", params)
        status = (data or {}).get("sendStatus") or {}
        return {
            "order_id": str(status.get("order_id") or ""),
            "closed_qty": qty,
            "realized_pnl_usd": float(match.get("unrealized_pnl_usd") or 0),
        }

    @classmethod
    async def list_positions(cls, creds: dict, symbol: str | None = None) -> list[dict]:
        data = await _signed_request(creds, "GET", "/openpositions")
        positions = (data or {}).get("openPositions") or []
        out: list[dict] = []
        for p in positions:
            pf_sym = p.get("symbol") or ""
            base = _denorm_symbol(pf_sym)
            if symbol and base.upper() != symbol.upper():
                continue
            try:
                qty = float(p.get("size") or 0)
            except (TypeError, ValueError):
                qty = 0.0
            if qty == 0:
                continue
            side = "buy" if (p.get("side") or "").lower() == "long" else "sell"
            try:
                funding = float(p.get("unrealizedFunding") or 0)
            except (TypeError, ValueError):
                funding = 0.0
            try:
                pnl = float(p.get("pnl") or 0)
            except (TypeError, ValueError):
                pnl = 0.0
            out.append({
                "exchange": "kraken",
                "symbol": base,
                "side": side,
                "quantity": abs(qty),
                "entry_price": float(p.get("price") or 0),
                "unrealized_pnl_usd": funding + pnl,
                "funding_pnl_usd": funding,  # paid funding (negative when user pays)
                "leverage": None,
                "margin_mode": "cross",
            })
        return out

    @classmethod
    async def get_public_qty_limits(cls, symbol: str) -> dict | None:
        """Kraken Futures expresses qty in CONTRACTS where 1 contract =
        `contract_size` coins. Min order = 1 contract. Position cap from
        instrument metadata (max_position_size in coins, post-multiply)."""
        try:
            info = (await _instruments()).get(_norm_symbol(symbol))
        except Exception:
            return None
        if not info:
            return None
        cs = float(info.get("contract_size") or 1)
        max_pos = float(info.get("max_position_size") or 0) or None
        return {
            "min_qty": cs,
            "step":    cs,
            "max_qty": max_pos,
            "unit": "coin",
        }

    @classmethod
    async def preflight(cls, creds, symbol, quantity, leverage):
        try:
            instr = await _instruments()
            sym = _norm_symbol(symbol)
            info = instr.get(sym)
            if not info:
                return {"ok": False, "reason": f"Kraken: no perp market for {symbol}"}
            cs = float(info.get("contract_size") or 1)
            if quantity < cs:
                return {"ok": False, "reason": f"Kraken min qty is {cs} {symbol.upper()} (1 contract)."}
            return {"ok": True}
        except Exception as exc:
            return {"ok": False, "reason": str(exc)[:180]}

File: go-fetcher/internal/trade/kraken/kraken.go
// Kraken Futures (futures.kraken.com /derivatives/api/v3) trade adapter.
//
// Port of `backend/services/trade_adapters/kraken.py`.
//
// Signing flavour (most exotic of the bunch):
//
//	hash    = SHA256(post_data + nonce + path)
//	authent = base64(HMAC_SHA512( base64_decode(api_secret), hash ))
//
// Headers:
//
//	APIKey:   <api_key>
//	Nonce:    <ms-timestamp>
//	Authent:  <base64 sig>
//
// Quirks:
//   - Symbol form: "PF_BTCUSD" (BTC mapped to XBT historic naming).
//   - Quantity in COINS (no contract conversion).
//   - One endpoint family: /derivatives/api/v3 prefixes the URL but
//     the SIGNATURE uses the path WITHOUT it (Kraken docs trap).
//   - SetLeverage hits /leveragepreferences; venue silently accepts
//     "already at this value" so we treat any error as soft.
package kraken

import (
	"context"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/bashyrov/wallet-monitor/go-fetcher/internal/trade"
)

const (
	baseURL = "https://futures.kraken.com"
	root    = "/derivatives/api/v3"
)

type Adapter struct {
	httpClient *http.Client
}

func New() *Adapter {
	return &Adapter{
		httpClient: &http.Client{
			Timeout: 15 * time.Second,
			Transport: &http.Transport{
				ForceAttemptHTTP2:   true,
				MaxIdleConns:        200,
				MaxIdleConnsPerHost: 32,
				MaxConnsPerHost:     64,
				IdleConnTimeout:     300 * time.Second,
				TLSHandshakeTimeout: 5 * time.Second,
			},
		},
	}
}

func init() { trade.Register("kraken", New()) }

func (a *Adapter) Name() string { return "kraken" }

// ── Symbol mapping ───────────────────────────────────────────────────────

func toKrakenSymbol(sym string) string {
	s := strings.ToUpper(sym)
	if s == "BTC" {
		s = "XBT"
	}
	return "PF_" + s + "USD"
}

func fromKrakenSymbol(pf string) string {
	s := strings.TrimPrefix(pf, "PF_")
	s = strings.TrimSuffix(s, "USD")
	if s == "XBT" {
		s = "BTC"
	}
	return s
}

// ── Signing ──────────────────────────────────────────────────────────────

func krakenSign(apiSecret, postData, nonce, path string) (string, error) {
	secret, err := base64.StdEncoding.DecodeString(apiSecret)
	if err != nil {
		return "", err
	}
	preHash := sha256.Sum256([]byte(postData + nonce + path))
	mac := trade.HMACWith(sha512.New, string(secret), string(preHash[:]))
	return base64.StdEncoding.EncodeToString(mac), nil
}

func (a *Adapter) signedRequest(
	ctx context.Context, creds trade.Creds, method, path string,
	params map[string]string,
) (json.RawMessage, error) {
	postData := ""
	if len(params) > 0 {
		postData = url.Values{}.Encode() // start empty
		v := url.Values{}
		for k, val := range params {
			v.Set(k, val)
		}
		postData = v.Encode()
	}
	nonce := strconv.FormatInt(time.Now().UnixMilli(), 10)
	authent, err := krakenSign(creds.APISecret, postData, nonce, path)
	if err != nil {
		return nil, errInternal("base64 secret decode", err)
	}

	u := baseURL + root + path
	var bodyReader io.Reader
	if method == http.MethodGet && postData != "" {
		u += "?" + postData
	} else if method != http.MethodGet && postData != "" {
		bodyReader = strings.NewReader(postData)
	}
	req, err := http.NewRequestWithContext(ctx, method, u, bodyReader)
	if err != nil {
		return nil, err
	}
	req.Header.Set("APIKey", creds.APIKey)
	req.Header.Set("Nonce", nonce)
	req.Header.Set("Authent", authent)
	req.Header.Set("Accept", "application/json")
	if method != http.MethodGet {
		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	}
	resp, err := a.httpClient.Do(req)
	if err != nil {
		return nil, &trade.Error{Kind: trade.KindTransient, Message: err.Error(), Cause: err}
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 400 {
		return nil, parseError(resp.StatusCode, raw)
	}
	var env struct {
		Result string `json:"result"`
		Error  string `json:"error"`
	}
	_ = json.Unmarshal(raw, &env)
	if env.Result == "error" {
		return nil, &trade.Error{Kind: trade.KindExchange, Message: env.Error}
	}
	return raw, nil
}

func parseError(status int, body []byte) *trade.Error {
	var env struct {
		Error string `json:"error"`
	}
	_ = json.Unmarshal(body, &env)
	msg := env.Error
	if msg == "" {
		msg = strings.TrimSpace(string(body))
	}
	if status == 429 {
		return &trade.Error{Kind: trade.KindRateLimit, Message: msg}
	}
	return &trade.Error{Kind: trade.KindExchange, Message: msg}
}

// ── Adapter methods ──────────────────────────────────────────────────────

func (a *Adapter) GetBalance(ctx context.Context, creds trade.Creds) (*trade.Balance, error) {
	body, err := a.signedRequest(ctx, creds, http.MethodGet, "/accounts", nil)
	if err != nil {
		return nil, err
	}
	var env struct {
		Accounts struct {
			Flex struct {
				BalanceValue json.Number `json:"balanceValue"`
			} `json:"flex"`
			Cash struct {
				Balances map[string]json.Number `json:"balances"`
			} `json:"cash"`
		} `json:"accounts"`
	}
	if err := json.Unmarshal(body, &env); err != nil {
		return nil, errInternal("parse balance", err)
	}
	usdt, _ := env.Accounts.Flex.BalanceValue.Float64()
	if usdt == 0 {
		usd, _ := env.Accounts.Cash.Balances["USD"].Float64()
		usdtBal, _ := env.Accounts.Cash.Balances["USDT"].Float64()
		usdt = usd + usdtBal
	}
	return &trade.Balance{TotalUSD: usdt, AvailableUSD: usdt}, nil
}

func (a *Adapter) SetLeverage(ctx context.Context, creds trade.Creds, req trade.LeverageRequest) error {
	if req.Leverage <= 0 {
		return errUser("leverage must be > 0")
	}
	_, err := a.signedRequest(ctx, creds, http.MethodPost,
		"/leveragepreferences", map[string]string{
			"symbol":      toKrakenSymbol(req.Symbol),
			"maxLeverage": strconv.Itoa(req.Leverage),
		})
	if err != nil {
		// Already-set / not-modified — Kraken returns various error
		// strings. Treat all set-leverage errors as soft (the order
		// can still proceed at whatever leverage is currently set).
		return nil
	}
	return nil
}

func (a *Adapter) PlaceOrder(ctx context.Context, creds trade.Creds, req trade.OpenRequest) (*trade.Result, error) {
	if err := req.Validate(); err != nil {
		return nil, err
	}
	side := "buy"
	if req.Side == trade.SideSell {
		side = "sell"
	}
	body, err := a.signedRequest(ctx, creds, http.MethodPost, "/sendorder",
		map[string]string{
			"orderType": "mkt",
			"symbol":    toKrakenSymbol(req.Symbol),
			"side":      side,
			"size":      qtyString(req.Quantity),
		})
	if err != nil {
		return nil, err
	}
	var resp struct {
		SendStatus struct {
			OrderID    string `json:"order_id"`
			FillStatus struct {
				Price json.Number `json:"price"`
			} `json:"fillStatus"`
		} `json:"sendStatus"`
	}
	_ = json.Unmarshal(body, &resp)
	avg, _ := resp.SendStatus.FillStatus.Price.Float64()
	return &trade.Result{
		OrderID:   resp.SendStatus.OrderID,
		Symbol:    req.Symbol,
		Side:      req.Side,
		Quantity:  req.Quantity,
		AvgPrice:  avg,
		Status:    "NEW",
		CreatedAt: time.Now().UTC(),
		Raw:       body,
	}, nil
}

func (a *Adapter) ClosePosition(ctx context.Context, creds trade.Creds, req trade.CloseRequest) (*trade.Result, error) {
	if req.Symbol == "" {
		return nil, errUser("symbol required")
	}
	positions, err := a.ListPositions(ctx, creds, req.Symbol)
	if err != nil {
		return nil, err
	}
	if len(positions) == 0 {
		return &trade.Result{Symbol: req.Symbol, Status: "FLAT"}, nil
	}
	p := positions[0]
	reduceSide := "sell"
	if p.Side == trade.SideSell {
		reduceSide = "buy"
	}
	body, err := a.signedRequest(ctx, creds, http.MethodPost, "/sendorder",
		map[string]string{
			"orderType":  "mkt",
			"symbol":     toKrakenSymbol(req.Symbol),
			"side":       reduceSide,
			"size":       qtyString(p.Quantity),
			"reduceOnly": "true",
		})
	if err != nil {
		return nil, err
	}
	var resp struct {
		SendStatus struct {
			OrderID string `json:"order_id"`
		} `json:"sendStatus"`
	}
	_ = json.Unmarshal(body, &resp)
	closeSide := trade.SideSell
	if reduceSide == "buy" {
		closeSide = trade.SideBuy
	}
	return &trade.Result{
		OrderID:   resp.SendStatus.OrderID,
		Symbol:    req.Symbol,
		Side:      closeSide,
		Quantity:  p.Quantity,
		Status:    "NEW",
		CreatedAt: time.Now().UTC(),
		Raw:       body,
	}, nil
}

func (a *Adapter) ListPositions(ctx context.Context, creds trade.Creds, symbol string) ([]trade.Position, error) {
	body, err := a.signedRequest(ctx, creds, http.MethodGet, "/openpositions", nil)
	if err != nil {
		return nil, err
	}
	var env struct {
		OpenPositions []struct {
			Symbol      string      `json:"symbol"`
			Side        string      `json:"side"` // long / short
			Size        json.Number `json:"size"`
			Price       json.Number `json:"price"` // entry
			MarkPrice   json.Number `json:"markPrice"`
			UnrealizedFunding json.Number `json:"unrealizedFunding"`
			MaxLeverage json.Number `json:"maxLeverage"`
		} `json:"openPositions"`
	}
	if err := json.Unmarshal(body, &env); err != nil {
		return nil, errInternal("parse positions", err)
	}
	wantSym := strings.ToUpper(symbol)
	out := make([]trade.Position, 0, len(env.OpenPositions))
	for _, p := range env.OpenPositions {
		base := fromKrakenSymbol(p.Symbol)
		if wantSym != "" && base != wantSym {
			continue
		}
		size, _ := p.Size.Float64()
		if size == 0 {
			continue
		}
		side := trade.SideBuy
		if strings.EqualFold(p.Side, "short") {
			side = trade.SideSell
		}
		entry, _ := p.Price.Float64()
		mark, _ := p.MarkPrice.Float64()
		fund, _ := p.UnrealizedFunding.Float64()
		lev, _ := p.MaxLeverage.Float64()
		out = append(out, trade.Position{
			Symbol:        base,
			Side:          side,
			Quantity:      math.Abs(size),
			EntryPrice:    entry,
			MarkPrice:     mark,
			Leverage:      int(lev),
			UnrealizedPnL: fund, // Kraken's flex collateral mode
			Notional:      math.Abs(size) * mark,
		})
	}
	return out, nil
}

// ── Helpers ──────────────────────────────────────────────────────────────

func qtyString(q float64) string {
	s := strconv.FormatFloat(q, 'f', 8, 64)
	if strings.Contains(s, ".") {
		s = strings.TrimRight(s, "0")
		s = strings.TrimRight(s, ".")
		if s == "" {
			s = "0"
		}
	}
	return s
}

func errUser(msg string, args ...any) *trade.Error {
	return &trade.Error{Kind: trade.KindUser, Message: fmt.Sprintf(msg, args...)}
}

func errInternal(msg string, cause error) *trade.Error {
	return &trade.Error{Kind: trade.KindInternal, Message: msg, Cause: cause}
}

var _ trade.Adapter = (*Adapter)(nil)
